lighthouse-core/report/v2/report-template.html

```diff
@@ -21,14 +21,15 @@
   <meta charset="utf-8">
   <meta name="viewport" content="width=device-width, initial-scale=1, minimum-scale=1">
   <title>Lighthouse Report</title>
+  <style>%%LIGHTHOUSE_CSS%%</style>
 </head>
 <body>
   <noscript>Lighthouse report requires JavaScript. Please enable.</noscript>
   <script>window.__LIGHTHOUSE_JSON__ = %%LIGHTHOUSE_JSON%%</script>
   <script>%%LIGHTHOUSE_JAVASCRIPT%%</script>
   <script>
-    const renderer = new ReportRenderer(window.__LIGHTHOUSE_JSON__);
-    renderer.render(document.body);
+    const renderer = new ReportRenderer(document);
+    document.body.appendChild(renderer.renderReport(window.__LIGHTHOUSE_JSON__));
```

Comment:
I don't think those goals are at odds with a declarative approach. For example, we could declare `<template>` partials in the page:

```
<template id="template-audit">
  <div class="lighthouse-audit">
    <div class="lighthouse-score">
      <div class="lighthouse-score__value"><!--score--></div>
      <div class="lighthouse-score__text">
        <div class="lighthouse-score__title"><!--title--></div>
        <div class="lighthouse-score__description"><!--detail--></div>
      </div>
    </div>
  </div>
</template>
```
http://jsbin.com/bojinipecu/edit?html,output

And various embedders could instantiate slightly different `<template>`s based on their needs. Alternatively, override a `render` method and add their extra functionality. It's still 100% DOM, but it would be easier to reason about app structure and make modifications to the UI.

This would also lend itself nicely to creating reusable chunks of markup+functionality in the form of web components. Each component would be responsible for its own css/html/js. DevTools could extend a given component if it needs something different.

```
const categories = __LIGHTHOUSE_JSON__.reportCategories;

const container = document.querySelector('lighthouse-category');
categories.forEach(audit => {
  const auditEl = document.createElement('lighthouse-audit');
  auditEl.audit = audit;
  container.appendChild(auditEl);
});
```

might produce something like:

```
<lighthouse-category>
  <lighthouse-audit id="critical-request-chains">
    #shadow-root
      <div class="lighthouse-score">...</div>
      ...
      <card-fortmatter></card-fortmatter>
  </lighthouse-audit>
  <lighthouse-audit>...</lighthouse-audit>
</lighthouse-category>
```